File: secret_protector/detector.py
from __future__ import annotations

import fnmatch
import re
from typing import Any

from .io_utils import eprint
from .policy import as_list, get_nested

PATH_LIKE_KEYS = {
    "path",
    "filepath",
    "file_path",
    "filename",
    "uri",
    "absolute_path",
    "absolutepath",
    "relative_path",
    "relativepath",
    "relative_workspace_path",
    "relativeworkspacepath",
}


def collect_strings(value: Any, out: list[str]) -> None:
    if isinstance(value, str):
        out.append(value)
        return

    if isinstance(value, dict):
        for key, item in value.items():
            out.append(str(key))
            collect_strings(item, out)
        return

    if isinstance(value, list):
        for item in value:
            collect_strings(item, out)


# We intentionally use loose heuristics here because tool payloads vary widely
# across providers and versions; path-like keys plus slash/dot prefixes catches
# most real payloads without strict schema coupling.
def collect_paths(value: Any, out: list[str], key_hint: str | None = None) -> None:
    if isinstance(value, dict):
        for key, item in value.items():
            normalized_key = re.sub(r"[^a-zA-Z0-9_]", "", str(key)).lower()
            collect_paths(item, out, normalized_key)
        return

    if isinstance(value, list):
        for item in value:
            collect_paths(item, out, key_hint)
        return

    if not isinstance(value, str):
        return

    is_path_key = key_hint in PATH_LIKE_KEYS if key_hint else False
    seems_like_path = "/" in value or "\\" in value or value.startswith(".")
    if is_path_key or seems_like_path:
        out.append(value)


def normalize_path(path: str) -> str:
    return path.strip().replace("\\", "/")


def compile_regexes(patterns: list[str]) -> list[re.Pattern[str]]:
    compiled: list[re.Pattern[str]] = []
    for pattern in patterns:
        try:
            compiled.append(re.compile(pattern))
        except re.error as exc:
            eprint(f"warning: invalid regex ignored: {pattern!r} ({exc})")
    return compiled


def first_env_match(text: str, env_exact: list[str], env_regex: list[re.Pattern[str]]) -> str | None:
    for name in env_exact:
        checks = [
            rf"\$\{{?{re.escape(name)}\}}?",
            rf"\b{re.escape(name)}\b\s*=",
            rf"\bexport\s+{re.escape(name)}\b",
            rf"\b{re.escape(name)}\b",
        ]
        for pattern in checks:
            if re.search(pattern, text):
                return name

    for pattern in env_regex:
        match = pattern.search(text)
        if match:
            return match.group(0)

    return None


def path_matches(path: str, file_globs: list[str], file_regex: list[re.Pattern[str]]) -> str | None:
    normalized = normalize_path(path)
    basename = normalized.split("/")[-1]

    for glob in file_globs:
        if fnmatch.fnmatch(normalized, glob) or fnmatch.fnmatch(basename, glob):
            return glob

    for pattern in file_regex:
        if pattern.search(normalized):
            return pattern.pattern

    return None


def policy_matchers(policy: dict[str, Any]) -> tuple[list[str], list[re.Pattern[str]], list[str], list[re.Pattern[str]]]:
    env_exact = [name.strip() for name in as_list(get_nested(policy, "env", "exact", default=[])) if name.strip()]
    env_regex = compile_regexes(as_list(get_nested(policy, "env", "regex", default=[])))

    file_globs = [glob.strip() for glob in as_list(get_nested(policy, "files", "globs", default=[])) if glob.strip()]
    file_regex = compile_regexes(as_list(get_nested(policy, "files", "regex", default=[])))
    return env_exact, env_regex, file_globs, file_regex


def detect_secret_leak(payload: Any, policy: dict[str, Any]) -> str | None:
    env_exact, env_regex, file_globs, file_regex = policy_matchers(policy)

    strings: list[str] = []
    collect_strings(payload, strings)
    for text in strings:
        env_hit = first_env_match(text, env_exact, env_regex)
        if env_hit:
            return f"Detected secret environment variable reference: {env_hit}"

    paths: list[str] = []
    collect_paths(payload, paths)
    for path in paths:
        path_hit = path_matches(path, file_globs, file_regex)
        if path_hit:
            return f"Detected sensitive file path pattern: {path_hit}"

    return None


def detect_sensitive_read(payload: Any, policy: dict[str, Any]) -> str | None:
    _, _, file_globs, file_regex = policy_matchers(policy)

    paths: list[str] = []
    collect_paths(payload, paths)
    for path in paths:
        hit = path_matches(path, file_globs, file_regex)
        if hit:
            return f"Read blocked for sensitive file pattern: {hit}"

    return None


def detect_sensitive_command(payload: Any, policy: dict[str, Any]) -> str | None:
    env_exact, env_regex, file_globs, file_regex = policy_matchers(policy)

    strings: list[str] = []
    collect_strings(payload, strings)
    for text in strings:
        env_hit = first_env_match(text, env_exact, env_regex)
        if env_hit:
            return f"Command references secret environment variable: {env_hit}"

    paths: list[str] = []
    collect_paths(payload, paths)
    for path in paths:
        path_hit = path_matches(path, file_globs, file_regex)
        if path_hit:
            return f"Command references sensitive file pattern: {path_hit}"

    return None

File: src/detector.ts
import * as path from "node:path";
import { minimatch } from "minimatch";
import { eprint } from "./io.js";
import { asList, getNested } from "./policy.js";

const BUILTIN_PATH_LIKE_KEYS = new Set([
  "path",
  "filepath",
  "file_path",
  "filename",
  "uri",
  "absolute_path",
  "absolutepath",
  "relative_path",
  "relativepath",
  "relative_workspace_path",
  "relativeworkspacepath",
]);

const BUILTIN_FILE_READ_COMMANDS = new Set([
  "cat",
  "head",
  "tail",
  "less",
  "more",
  "grep",
  "awk",
  "sed",
  "bat",
  "rg",
]);

export function getPathLikeKeys(policy: Record<string, unknown>): Set<string> {
  const configured = asList(getNested(policy, "detection", "path_like_keys") ?? []);
  if (configured.length === 0) return BUILTIN_PATH_LIKE_KEYS;
  return new Set([...BUILTIN_PATH_LIKE_KEYS, ...configured.map((k) => String(k).replace(/[^a-zA-Z0-9_]/g, "").toLowerCase())]);
}

export function getFileReadCommands(policy: Record<string, unknown>): Set<string> {
  const configured = asList(getNested(policy, "detection", "file_read_commands") ?? []);
  if (configured.length === 0) return BUILTIN_FILE_READ_COMMANDS;
  return new Set([...BUILTIN_FILE_READ_COMMANDS, ...configured.map((k) => String(k).toLowerCase())]);
}

export function collectStrings(value: unknown, out: string[]): void {
  if (typeof value === "string") {
    out.push(value);
    return;
  }
  if (typeof value === "object" && value !== null) {
    if (Array.isArray(value)) {
      for (const item of value) collectStrings(item, out);
    } else {
      for (const [key, item] of Object.entries(value)) {
        out.push(String(key));
        collectStrings(item, out);
      }
    }
  }
}

export function collectPaths(
  value: unknown,
  out: string[],
  keyHint?: string,
  pathLikeKeys?: Set<string>
): void {
  const keys = pathLikeKeys ?? BUILTIN_PATH_LIKE_KEYS;
  if (typeof value === "object" && value !== null) {
    if (Array.isArray(value)) {
      for (const item of value) collectPaths(item, out, keyHint, keys);
      return;
    }
    for (const [key, item] of Object.entries(value)) {
      const normalizedKey = String(key).replace(/[^a-zA-Z0-9_]/g, "").toLowerCase();
      collectPaths(item, out, normalizedKey, keys);
    }
    return;
  }
  if (typeof value !== "string") return;
  const isPathKey = keyHint ? keys.has(keyHint) : false;
  const seemsLikePath = value.includes("/") || value.includes("\\") || value.startsWith(".");
  if (isPathKey || seemsLikePath) out.push(value);
}

export function normalizePath(p: string): string {
  return p.trim().replace(/\\/g, "/");
}

function compileRegexes(patterns: string[]): RegExp[] {
  const compiled: RegExp[] = [];
  for (let pattern of patterns) {
    const iFlag = /^\(\?i\)/.test(pattern);
    if (iFlag) pattern = pattern.replace(/^\(\?i\)/, "");
    try {
      compiled.push(new RegExp(pattern, iFlag ? "i" : ""));
    } catch (err) {
      eprint(`warning: invalid regex ignored: ${JSON.stringify(pattern)} (${err})`);
    }
  }
  return compiled;
}

function getEnvReferencePatterns(policy: Record<string, unknown>): string[] {
  return asList(getNested(policy, "detection", "env_reference_patterns") ?? []);
}

export function firstEnvMatch(
  text: string,
  envExact: string[],
  envRegex: RegExp[],
  envReferencePatterns?: string[]
): string | null {
  for (const name of envExact) {
    const escaped = name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const checks: RegExp[] = [
      new RegExp(`\\$\\{?${escaped}\\}?`),
      new RegExp(`\\b${escaped}\\b\\s*=`),
      new RegExp(`\\bexport\\s+${escaped}\\b`),
      new RegExp(`\\b${escaped}\\b`),
    ];
    if (envReferencePatterns?.length) {
      for (const tpl of envReferencePatterns) {
        try {
          const pattern = tpl.replace(/\{NAME\}/g, escaped);
          checks.push(new RegExp(pattern));
        } catch {
          // skip invalid pattern
        }
      }
    }
    for (const re of checks) {
      if (re.test(text)) return name;
    }
  }
  for (const re of envRegex) {
    const m = text.match(re);
    if (m) return m[0];
  }
  return null;
}

/**
 * Best-effort extraction of file paths from a shell command string.
 * Handles: cat .env, grep x .env.local, head -n 1 .env, rg pattern -- .env.
 * Limitations: subshells $(...), backticks, variable expansion in paths not parsed.
 */
export function extractPathsFromCommand(
  command: string,
  fileReadCommands?: Set<string>
): string[] {
  if (typeof command !== "string" || !command.trim()) return [];

  const commands = fileReadCommands ?? BUILTIN_FILE_READ_COMMANDS;
  const tokens = tokenizeCommand(command);
  const paths: string[] = [];
  let afterFileReadCommand = false;
  let skipUntilNextArg = false;

  for (const t of tokens) {
    if (t === "|" || t === ";" || t === "&" || t === "&&" || t === "||") {
      afterFileReadCommand = false;
      skipUntilNextArg = false;
      continue;
    }

    if (commands.has(t.toLowerCase())) {
      afterFileReadCommand = true;
      skipUntilNextArg = false;
      continue;
    }

    if (t === "--") {
      skipUntilNextArg = false;
      afterFileReadCommand = true;
      continue;
    }

    if (t.startsWith("-") && t !== "--") {
      if (afterFileReadCommand) skipUntilNextArg = true;
      continue;
    }

    if (skipUntilNextArg && afterFileReadCommand) {
      skipUntilNextArg = false;
      continue;
    }

    if (afterFileReadCommand) {
      if (t.includes("/") || t.includes("\\") || t.startsWith(".") || /^[\w.-]+$/.test(t) || t.includes(" ")) {
        paths.push(t);
      }
    }
  }
  return paths;
}

function tokenizeCommand(command: string): string[] {
  const tokens: string[] = [];
  let i = 0;
  while (i < command.length) {
    while (i < command.length && /\s/.test(command[i])) i++;
    if (i >= command.length) break;

    const ch = command[i];
    if (ch === '"' || ch === "'") {
      const quote = ch;
      i++;
      let end = i;
      while (end < command.length && command[end] !== quote) {
        if (command[end] === "\\") end++;
        end++;
      }
      tokens.push(command.slice(i, end).replace(/\\"/g, '"'));
      i = end < command.length ? end + 1 : end;
    } else {
      let end = i;
      while (end < command.length && !/\s/.test(command[end]) && command[end] !== "|" && command[end] !== ";" && command[end] !== "&" && command[end] !== '"' && command[end] !== "'") {
        if (command[end] === "\\") end++;
        end++;
      }
      tokens.push(end > i ? command.slice(i, end) : command[i] ?? "");
      i = end === i && end < command.length ? end + 1 : end;
    }
  }
  return tokens;
}

export function pathMatches(
  p: string,
  fileGlobs: string[],
  fileRegex: RegExp[]
): string | null {
  const normalized = normalizePath(p);
  const basename = path.basename(normalized);

  for (const glob of fileGlobs) {
    try {
      if (minimatch(normalized, glob) || minimatch(basename, glob)) return glob;
    } catch {
      // skip invalid glob
    }
  }
  for (const re of fileRegex) {
    if (re.test(normalized)) return re.source;
  }
  return null;
}

function pathMatchesAllow(
  p: string,
  allowGlobs: string[],
  allowRegex: RegExp[]
): boolean {
  const normalized = normalizePath(p);
  const basename = path.basename(normalized);
  for (const glob of allowGlobs) {
    try {
      if (minimatch(normalized, glob) || minimatch(basename, glob)) return true;
    } catch {
      // skip invalid glob
    }
  }
  for (const re of allowRegex) {
    if (re.test(normalized)) return true;
  }
  return false;
}

export function pathMatchesWithAllow(
  p: string,
  blockGlobs: string[],
  blockRegex: RegExp[],
  allowGlobs: string[],
  allowRegex: RegExp[]
): string | null {
  if (pathMatchesAllow(p, allowGlobs, allowRegex)) return null;
  return pathMatches(p, blockGlobs, blockRegex);
}

export function policyMatchers(
  policy: Record<string, unknown>
): [string[], RegExp[], string[], RegExp[], string[], RegExp[]] {
  const envExact = asList(getNested(policy, "env", "block_exact") ?? []).filter((s) => s.trim());
  const envRegex = compileRegexes(asList(getNested(policy, "env", "block_regex") ?? []));
  const fileGlobs = asList(getNested(policy, "files", "block_globs") ?? []).filter((s) => s.trim());
  const fileRegex = compileRegexes(asList(getNested(policy, "files", "block_regex") ?? []));
  const allowGlobs = asList(getNested(policy, "files", "allow_globs") ?? []).filter((s) => s.trim());
  const allowRegex = compileRegexes(asList(getNested(policy, "files", "allow_regex") ?? []));
  return [envExact, envRegex, fileGlobs, fileRegex, allowGlobs, allowRegex];
}

export function detectSecretLeak(
  payload: unknown,
  policy: Record<string, unknown>
): string | null {
  const [envExact, envRegex, fileGlobs, fileRegex, allowGlobs, allowRegex] = policyMatchers(policy);
  const pathLikeKeys = getPathLikeKeys(policy);
  const envRefPatterns = getEnvReferencePatterns(policy);

  const strings: string[] = [];
  collectStrings(payload, strings);
  for (const text of strings) {
    const envHit = firstEnvMatch(text, envExact, envRegex, envRefPatterns.length ? envRefPatterns : undefined);
    if (envHit) return `Detected secret environment variable reference: ${envHit}`;
  }

  const paths: string[] = [];
  collectPaths(payload, paths, undefined, pathLikeKeys);
  for (const p of paths) {
    const hit = pathMatchesWithAllow(p, fileGlobs, fileRegex, allowGlobs, allowRegex);
    if (hit) return `Detected sensitive file path pattern: ${hit}`;
  }
  return null;
}

export function detectSensitiveRead(
  payload: unknown,
  policy: Record<string, unknown>
): string | null {
  const [, , fileGlobs, fileRegex, allowGlobs, allowRegex] = policyMatchers(policy);
  const pathLikeKeys = getPathLikeKeys(policy);
  const paths: string[] = [];
  collectPaths(payload, paths, undefined, pathLikeKeys);
  for (const p of paths) {
    const hit = pathMatchesWithAllow(p, fileGlobs, fileRegex, allowGlobs, allowRegex);
    if (hit) return `Read blocked for sensitive file pattern: ${hit}`;
  }
  return null;
}

export function detectSensitiveCommand(
  payload: unknown,
  policy: Record<string, unknown>
): string | null {
  const [envExact, envRegex, fileGlobs, fileRegex, allowGlobs, allowRegex] = policyMatchers(policy);
  const pathLikeKeys = getPathLikeKeys(policy);
  const fileReadCommands = getFileReadCommands(policy);

  const envRefPatterns = getEnvReferencePatterns(policy);
  const strings: string[] = [];
  collectStrings(payload, strings);
  for (const text of strings) {
    const envHit = firstEnvMatch(text, envExact, envRegex, envRefPatterns.length ? envRefPatterns : undefined);
    if (envHit) return `Command references secret environment variable: ${envHit}`;
  }

  const paths: string[] = [];
  collectPaths(payload, paths, undefined, pathLikeKeys);

  for (const text of strings) {
    const cmdPaths = extractPathsFromCommand(text, fileReadCommands);
    paths.push(...cmdPaths);
  }

  for (const p of paths) {
    const hit = pathMatchesWithAllow(p, fileGlobs, fileRegex, allowGlobs, allowRegex);
    if (hit) return `Command references sensitive file path pattern: ${hit}`;
  }
  return null;
}
